Feature - Report Creation Screen

Amerilla opened the pull request:
I created a basic implementation of the report creation screen in `ui/farmer/AddReportScreen.kt `and its associatd ViewModel in `ui/farmer/AddReportViewModel.kt`. We can add a title, description, and choose a vet (among a dummy list) for the report. The button that adds a picture from gallery/camera does not yet work.

The tests are part of a separate PR (but yet still briefly appear in this PR as a result of bad git manipulation)

MorgadoKevin commented on AgriHealth-Alert-main/app/src/main/java/com/android/agrihealth/ui/farmer/AddReportScreen.kt:
```diff
@@ -1,37 +1,278 @@
-package com.android.agrihealth.ui.farmer
+package com.android.agrihealth.ui.report
```
The package directive does not match file location.

MorgadoKevin commented on AgriHealth-Alert-main/app/src/main/java/com/android/agrihealth/ui/farmer/AddReportScreen.kt:
```diff
@@ -1,37 +1,278 @@
+import androidx.compose.ui.unit.dp
+import androidx.lifecycle.viewmodel.compose.viewModel
+import androidx.compose.foundation.shape.RoundedCornerShape
+import androidx.compose.material3.*
```
the imports contain material3 wildcard as well as individual material3 imports which is redundant and not needed.

MorgadoKevin commented on AgriHealth-Alert-main/app/src/main/java/com/android/agrihealth/ui/farmer/AddReportScreen.kt:
```diff
@@ -1,37 +1,278 @@
+                    }
+                },
+                navigationIcon = {
+                    IconButton(onClick = { onBack }) {
```
The icon button should have the test tag: modifier = Modifier.testTag(NavigationTestTags.GO_BACK_BUTTON))

MorgadoKevin commented on AgriHealth-Alert-main/app/src/main/java/com/android/agrihealth/ui/farmer/AddReportScreen.kt:
```diff
@@ -1,37 +1,278 @@
+                modifier =
+                    Modifier.fillMaxWidth()
+                        .height(56.dp)
+                        .testTag(SignUpScreenTestTags.SAVE_BUTTON),
```
Calls the wrong test tag. Should call AddReportScreenTestTag.CREATE_BUTTON

MorgadoKevin commented on AgriHealth-Alert-main/app/src/main/java/com/android/agrihealth/ui/farmer/AddReportScreen.kt:
```diff
@@ -1,37 +1,278 @@
+                        .height(56.dp)
+                        .testTag(SignUpScreenTestTags.SAVE_BUTTON),
+                shape = RoundedCornerShape(20.dp),
+                colors = ButtonDefaults.buttonColors(containerColor = Color(0xFF96B7B1))
```
The colors could be declared at the top, similar to how it is done for Field.

Amerilla commented:
I addressed the all of the issues found in the PR review. I then ran `ktfmtFormat` and made some minor adjustments to the tests.

I apologize for the messy commit history, my IDE was corrupt and failed to function correctly. My attempt to fix the situation resulted in that disorganized history

al23029 commented on AgriHealth-Alert-main/app/src/main/java/com/android/agrihealth/ui/farmer/AddReportScreen.kt:
```diff
@@ -1,37 +1,278 @@
+            verticalArrangement = Arrangement.spacedBy(16.dp)
+        )
+        {
+            Spacer(Modifier.height(96.dp))
```
This applies not only to the specific place he commented on but to the entire screen.
When you specify a fixed height like .height(56.dp), the layout won't automatically adjust for different screen sizes.Please try to use contentAlignment = Alignment.Center or similar approaches wherever possible to make the layout more flexible.

al23029 commented on AgriHealth-Alert-main/app/src/main/java/com/android/agrihealth/ui/farmer/AddReportScreen.kt:
```diff
@@ -1,41 +1,223 @@
+}
+
+private val unfocusedFieldColor = Color(0xFFF0F7F1)
+private val focusedFieldColor = Color(0xFFF0F7F1)
```
I noticed the first two colors are currently the same. Do we need to keep both definitions, or are you planning to differentiate them later?

al23029 commented on AgriHealth-Alert-main/app/src/main/java/com/android/agrihealth/ui/farmer/AddReportScreen.kt:
```diff
@@ -1,41 +1,223 @@
+              Spacer(Modifier.height(28.dp))
+              Button(
+                  onClick = {
+                    val created = createReportViewModel.createReport()
```
Consider making createReport() a suspend function and using withContext(Dispatchers.IO) for the repository call. This way, the coroutine will wait until the report is actually added before returning, which can prevent race conditions or false success states.
Maybe this is the cause of why, when testing the app, the created report doesn’t appear in the Overview screen yet.

Amerilla replied:
Now this is done. The report is indeed added to the repository, maybe the issue is in the Overview screen?

al23029 commented on AgriHealth-Alert-main/app/src/main/java/com/android/agrihealth/ui/farmer/AddReportViewModel.kt:
```diff
@@ -0,0 +1,68 @@
+    viewModelScope.launch { reportRepository.addReport(newReport) }
+
+    // Clears all the fields
+    clearInputs()
```
clearInputs() is called regardless of whether report creation succeeds or fails, but it would be more user-friendly to keep the inputs when creation fails.

Amerilla replied:
Currenty `reportRepository.addReport(newReport)` has no return type so there is no way to verify that adding the report to the repository was successful. I think we should also modify the repository so that it itself checks that the record is valid and return a success status

al23029 reviewed (changes requested):
When testing on a real device, I noticed two things:
- After successfully creating a report, it would be more user-friendly if the screen automatically navigated back to the Overview screen.
- Even after creating a report, it doesn’t seem to appear in the Overview screen yet.

I also left a few additional comments for things that caught my attention, please go through them one by one and REPLY to each to confirm that you addressed it.
I went through Kevin’s comments as well and marked the resolved ones with 👍, but one issue remained unresolved, so I reopened it and added some clarification.

I’ve fetched this PR locally, confirmed that all tests pass, and that the coverage report shows over 80%.
However, since an important CI-related PR was merged after your last push, please pull the latest changes before pushing again.
This will ensure that CI runs correctly and the code coverage check passes on the PR.

MorgadoKevin commented on AgriHealth-Alert-main/app/src/main/java/com/android/agrihealth/ui/farmer/AddReportScreen.kt:
```diff
@@ -1,41 +1,223 @@
+                    scope.launch {
+                      if (created) {
+                        snackbarHostState.showSnackbar(AddReportFeedbackTexts.SUCCESS)
+                        onCreateReport()
```
the code seems to be waiting for the snackbar to disappear before calling onCreateReport() which is confusing for the user. Consider showing the snackbar on the overview screen if create report is successful. Or showing the snack bar for less time.

Amerilla replied:
I think having the snackbar appear on Overview Screen would require modifying the Overview Screen code. To keep it self-contained, I opted for a different approach: the snackbar is displayed only in case of failure (currently the addReport makes the check to see if a report is valid but the repository itself should do this and return a success status in my opinion), and a dialog is displayed in case of success. Once the dialog is dismissed, it goes automatically back to the overview screen

MorgadoKevin commented on AgriHealth-Alert-main/app/src/main/java/com/android/agrihealth/ui/farmer/AddReportViewModel.kt:
```diff
@@ -0,0 +1,68 @@
+            title = uiState.title,
+            description = uiState.description,
+            photoUri = null, // currently unused
+            farmerId = "currentUserId", //
```
Should add a TODO comment to use the current use id when implemented.

Amerilla replied:
I added the TODO

Amerilla commented on AgriHealth-Alert-main/app/src/main/java/com/android/agrihealth/ui/farmer/AddReportScreen.kt:
```diff
@@ -1,41 +1,223 @@
-  const val SCREEN = "AddReportScreen"
+  const val TITLE_FIELD = "titleField"
+  const val DESCRIPTION_FIELD = "descriptionField"
+  const val IMAGE_BUTTON = "imageButton"
```
Currently there is no add image feature implemented so that tag is indeed unnecessary for now. I removed it
